src/sentry/migrations/0348_issue_alert_fallback.py

```diff
@@ -0,0 +1,105 @@
+# Generated by Django 2.2.28 on 2023-01-04 19:14
+
+import logging
+from enum import Enum
+
+from django.db import migrations, transaction
+
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+
+
+# Redefining the enums here so that we can reliably use them in the migration.
+class RuleStatus(Enum):
+    ACTIVE = 0
+
+
+class ProjectStatus(Enum):
+    ACTIVE = 0
+    DISABLED = 1
+
+
+class OrganizationStatus(Enum):
+    ACTIVE = 0
+
+
+def set_issue_alert_fallback(rule, fallthrough_choice):
+    actions = rule.data.get("actions", [])
+    rule_changed = False
+    for action in actions:
+        id = action.get("id")
+        target_type = action.get("targetType")
+        if id == "sentry.mail.actions.NotifyEmailAction" and target_type == "IssueOwners":
+            action.update({"fallthroughType": fallthrough_choice})
+            rule_changed = True
+
+    if rule_changed:
+        rule.data["actions"] = actions
+        rule.save()
+
+
+def migrate_project_ownership_to_issue_alert_fallback(project, ProjectOwnership, Rule):
+    with transaction.atomic():
+        # Determine whether this project has a fallback setting.
+        fallthrough_choice = None
+        try:
+            ownership = ProjectOwnership.objects.get(project=project)
+            fallthrough_choice = "AllMembers" if ownership and ownership.fallthrough else "NoOne"
+        except ProjectOwnership.DoesNotExist:
+            # Projects without fallbacks will be assigned the new "ActiveMembers" default.
+            fallthrough_choice = "ActiveMembers"
+
+        # We only migrate rules that are not pending deletion.
+        for rule in Rule.objects.filter(project=project, status=RuleStatus.ACTIVE.value):
+            set_issue_alert_fallback(rule, fallthrough_choice)
+
+        project.save()
+
+
+def migrate_to_issue_alert_fallback(apps, schema_editor):
+    Project = apps.get_model("sentry", "Project")
+    ProjectOwnership = apps.get_model("sentry", "ProjectOwnership")
+    Organization = apps.get_model("sentry", "Organization")
+    Rule = apps.get_model("sentry", "Rule")
+
+    # We migrate a project at a time, but we prefer to group by org so that for the
+    # most part an org will see the changes all at once.
+    for org in RangeQuerySetWrapperWithProgressBar(
+        Organization.objects.filter(status=OrganizationStatus.ACTIVE.value)
+    ):
+        # We only migrate projects that are not pending deletion.
+        for project in Project.objects.filter(
+            organization=org, status__in=[ProjectStatus.ACTIVE.value, ProjectStatus.DISABLED.value]
+        ):
```

Comment:
Should this use RangeQuerySetWrapper as well? Some organizations have 1000+ projects.